### PR TITLE
Add support for SSO login authentication

### PR DESCRIPTION
**Pull request by Mikescops**
Relates to #103

### CODE REVIEW

File: src/modules/auth/sso/index.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                await page.goto(serviceProviderURL);
+            } catch (error) {
+                reject(error);
+            }
+        })();
+    });
+};
+
+export const doSSOVerification = async ({ requestedLogin, serviceProviderURL }: SSOParams) => {
+    const redirectURL = await openIdPAndWaitForRedirectURL(serviceProviderURL);
```
Comment (showi) [inline src/modules/auth/sso/index.ts:37]: So this will wait till end of time if there's no match.

Comment (Mikescops) [reply src/modules/auth/sso/index.ts:37]: that's a valid comment, I'll probably wait for the tab to be closed or a long timeout

File: src/modules/crypto/buildSsoRemoteKey.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    const remoteKey = remoteKeys.filter((key) => key.type === 'sso')[0];
+
+    if (!remoteKey) {
+        throw new Error('Remote SSO key is missing');
+    }
+
+    const ssoKeysXored = xor(Buffer.from(ssoServerKey, 'base64'), Buffer.from(ssoSpKey, 'base64'));
+
+    const buffer = Buffer.from(remoteKey.key, 'base64');
```
Comment (showi) [inline src/modules/crypto/buildSsoRemoteKey.ts:33]: should be remoteKeyBase64?

File: src/modules/crypto/decrypt.ts [hunk 3]
```diff
@@ -39,17 +63,27 @@ export const decrypt = async (encryptedAsBase64: string, symmetricKeyGetter: Sym
             let symmetricKeyPromise = symmetricKeyGetter.derivates.get(derivationMethodBytes);
             if (!symmetricKeyPromise) {
                 winston.debug(`Computing new derivate with method: ${encryptedData.keyDerivation.algo}`);
-                symmetricKeyPromise = getDerivateUsingParametersFromEncryptedData(
-                    symmetricKeyGetter.secrets.masterPassword,
-                    encryptedData
-                );
+                if (encryptedData.keyDerivation.algo === 'noderivation') {
```
Comment (showi) [inline src/modules/crypto/decrypt.ts:66]: why don't you treat this case in `getDerivateUsingParametersFromEncryptedData`

Comment (Mikescops) [reply src/modules/crypto/decrypt.ts:66]: sounds obvious now that you say it

File: src/modules/crypto/encryptedDataDeserialization.ts [hunk 4]
```diff
@@ -128,7 +128,7 @@ const deserializeSymmetricCipherConfig = (
 
     const mode = extractNextEncryptedDataStringComponent(encryptedDataString.substring(cursor));
     cursor += mode.cursorAfter;
-    if (mode.component !== 'cbchmac') {
+    if (mode.component !== 'cbchmac' && mode.component !== 'cbchmac64') {
```
Comment (showi) [inline src/modules/crypto/encryptedDataDeserialization.ts:131]: I don't like this form but that's just me

`if (!(mode.component === 'cbchmac' || mode.component === 'cbchmac64'))`

Comment (Mikescops) [reply src/modules/crypto/encryptedDataDeserialization.ts:131]: same stuff

File: src/modules/crypto/xor.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+/** XOR two keys together, expects buffers to be of the same length */
+export const xor = (leftKey: Buffer, rightKey: Buffer): Buffer => {
+    if (leftKey.length !== rightKey.length) {
```
Comment (showi) [inline src/modules/crypto/xor.ts:3]: can check the length here

`return Buffer.from(b1.map((b1, index) => b1 ^ b2[index]))`

**Review by showi — APPROVED** [review]

File: src/modules/crypto/xor.ts (hunk 5 above)
Comment (Mikescops) [reply src/modules/crypto/xor.ts:3]: i'd like to keep it easy to read

**Pull request merged**